### PR TITLE
DevWs for Vision-Elec. Technology Co., Ltd. containing containing Vision 4-in-1 sensor

### PR DESCRIPTION
**Pull request by vision-raytseng**

### CODE REVIEW

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Please change deviceJoinName to "Vision Multipurpose Sensor" in this fingerprint and add Raw Description in a comment after the fingerprint so it would look like:
fingerprint mfr:"0109", prod:"2021", model:"2112", deviceJoinName: "Vision Multipurpose Sensor" // Raw Description // Raw Description >>paste the raw description here<<

(Raw description can be found at the Smartthings graphapi in device details)

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy:6]: Please reformat the whole code and fix indentation (change 4 spaces to 1 tab) - applies to a whole file.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Please remove the whole tiles section since tiles are not supported in the SmartThings app.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy:219]: CommandClassVersions are not obligatory. You may use zwave.parse(description) instead.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Please remove this empty method.

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: What about the Raw Description in a comment after the fingerprint ?

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: What about the Raw Description in a comment after the fingerprint ?

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Could You remove this map since it's not used ?

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy:245]: Could You explain why do You call configure() on every WakeUpNotification ?

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Please use this convention of formatting (applies to all occurrences)
`} else if() {
...
}`

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Please remove all those empty lines (those ones after break statements) in this switch block.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Since second2Hour() is used only once, can You just inline this code here ?

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy:245]: This is battery-powered device that only wakes up and communicates when an event has occurred.
If the smartThings hub would like to communicate with this device, the smartThings hub must wait for this device awake. When this device wakes up, it will send WakeUpNotification to inform the smartThings hub it is ready to communicate, and then the smartThings hub could send the configuration from the user settings to this device. So I call the configure() on every receiving WakeUpNotification to check that command the user configured needs to be sent to this device.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Okay, I modified it.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: OKay, I have removed.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy:245]: That way configure() will be called regardless changes in settings and every few hours the device will receive the set of commands even if user doesn't intent to configure the device. You can improve that approach by using another state variables to determine if the configuration has to be done or not (for example: "Z-Wave Motion Sensor")

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Please add "Raw description" (literally), so it would look:
...deviceJoinName: "Vision Multipurpose Sensor" // Raw description: zw:Ss2a type...

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy:245]: Actually, the device does not receive the set of commands every few hours if the settings have not changed, since the configure() checks which configuration was changed by the user and then sends the command to the device. You could check the code of updated() and configure(). In updated(), it will set the corresponding flag if user changes in settings, and then in configure() it will check the flag and decide which command needs to be sent to the device. This mechanism is the same as you said.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Okay, I have added "Raw description" comment in fingerprint.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy:219]: I do not modify it, since I would like to specify the command class version for used.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (KKlimczukS) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Please, could You fix indentations in parameter7options and parameter7enumMap ?

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Okay, I will fix the indentations in parameter7options and parameter7enumMap.

Comment (KKlimczukS) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Looks better now, but now spacing has to be fixed (replace 4 spaces to 1 tab).

**Review by KKlimczukS — CHANGES_REQUESTED** [review]

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: These was used 8 tabs and 3 spaces to make these indentations originally before I fixed today, but it looks not alignment. So I replace the 7 tabs to 28 spaces to fix this look. If I replace 4 spaces to 1 tab, it will be the same as previous committed code and look not good.

Comment (KKlimczukS) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Since indentation in this file contains tabs - please use tabs in this case too (it doesn't have to be super aligned with the first element)

Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: So these indentations are allowed?
```
	parameter.parameter7options = ["LED off when Temperature report/Motion active",
		"LED blink when Temperature report/Motion active",
		"LED blink when Motion active/LED off when Temperature report"]
	parameter.parameter7enumMap = ["LED off when Temperature report/Motion active": 1,
		"LED blink when Temperature report/Motion active": 2,
		"LED blink when Motion active/LED off when Temperature report": 3]
```

Comment (KKlimczukS) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Yes, it would look better this way.

Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: I got it, thank you.

**Comment by greens** [comment]
@vision-raytseng just checking, other than preferences/configuration, what parts of the handler code for this device would not fit in https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/smartthings/zooz-4-in-1-sensor.src/zooz-4-in-1-sensor.groovy ?

**Comment by vision-raytseng** [comment]
@greens other than preferences and configurations, some differences are list as followed:
1. This DTH handles the commands of wakeUpIntervalSet and wakeUpIntervalReport.
2. This DTH does not send batteryGet command on every receiving wakeUpNotification, but it sends batteryGet command if it is more than 8 hours from receiving last battery report.
3. The temperature unit is according to the user's settings in this DTH, not according to location.
4. This DTH does not use basicSet command to handle motion event, just use notification command to handle motion event.
5. This DTH supports two secure encapsulation commands (secure v1 and secure v2), and which secure encapsulation command was used depends on adding to smartThings hub by which secure level. So this DTH does not specify any secure encapsulation command, and to use which secure encapsulation command to communicate with this device is decided by smartThings hub.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (greens) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: wouldn't it make more sense to do all of this as a map of maps? Would certainly make iteration etc a lot simpler for you.

i.e.: 

```
def parameter = [:]
parameter[1] = {name: "TemperatureUnit", "title": "Temperature Unit [C/F]" ..}
```

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (greens) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: you do not need to include all of these, only the ones whose messages you are parsing/handling

Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Okay, I fixed it. Now I only include these command version that I parse and handle.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: I think not. Using a map of maps does not change the iteration way in my case, since I still use the index to make iteration to get key and value. In my case, I use the parameter name as a part of key name of state (i.e. state."${getConfigurationInfo(idx, "name")}Refresh" and state."${getConfigurationInfo(idx, "name")}") to store which configuration is changed by user and check which configuration command needs to be sent.
I think a map of maps that maybe looks better but here it does not make iteration more simpler for me.
I paste the snippet code of this function using a map of maps.
```
def getConfigurationInfo(num, text) {
	def parameter = [:]
	parameter[1] = [
		name: "TemperatureUnit", 
		title: "Temperature Unit [°C/°F]",
		description: "", 
		type: "enum", 
		options: ["°C", "°F"], 
		enumMap: ["°C": 0, "°F": 1], 
		default: "°C", 
		size: 1
	]
	...
	parameter[8] = [
		name: "RetryTimes", 
		title: "Motion notification retry times", 
		description: "",
		type: "number", 
		options: "0..10", 
		enumMap: [], 
		default: 3, 
		size: 1
	]
	return parameter[num.toInteger()]."${text}"
}
```

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (greens) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: here and elsewhere we've provided as a convenience the `scaledConfigurationValue` where this conversion has already been done

conversely you can pass integers directly in configuration sets as well

you should be able to do away with your conversion functions

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (greens) [inline devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: if you wrote your configuration report hander function differently and refactored your configuration parameters as a map of maps, I think this method could instead be written:

```
config.each { state[key].Refresh = false}
```

where `state[x]` is a map containing a value with key `value` and a boolean with key `Refresh` and your configuration report handler returns a list of numbers.

Because as far as I can tell you're not actually comparing the value you've received from the configuration report to the value the user has set. Any configuration report indicating a parameter number the user has set will get marked as not needing to be refreshed, regardless of the actual value reported.

This is just one example here. I'm not really satisfied with how you've handled configuration options. For a more acceptable example take a look at any of the Fibaro DTHs that include preferences.

Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: > Because as far as I can tell you're not actually comparing the value you've received from the configuration report to the value the user has set. Any configuration report indicating a parameter number the user has set will get marked as not needing to be refreshed, regardless of the actual value reported.

Actually, I will compare the value I am received from the configuration report to the value the user has set. If these values are different, I will send the configuration value the user has set on next time I received the wake up notification until these value are the same. 

Here is my configuration snippet code, and its procedure. Please look this, and give me a suggestion, thanks.

Variable definition:
1. _**state."${getConfigurationInfo(idx, "name")}"**_: The values the user has set.
2. _**state."${getConfigurationInfo(idx,"name")}Refresh"**_:  This configuration is needed to be sent or not: true is needed to be sent while I received the wake up notification, false is not.

Procedure:
1. After the user changed the configuration value, I will check which configuration value is changed. If this configuration value has changed, I will store this configuration value to corresponding _**state."${getConfigurationInfo(idx, "name")}"**_, and set _**state."${getConfigurationInfo(idx, "name")}Refresh"**_ to true to indicate this needs to be updated to this device. Here are **updated()** and **configurationUpdate(settings)**.
2. I will send this configuration set with _**state."${getConfigurationInfo(idx, "name")}"**_ if its _**state."${getConfigurationInfo(idx, "name")}Refresh"**_ is true after I received the wake up notification. And then I will send the configuration get to check the configuration report value is the same or not. Here are z**waveEvent(physicalgraph.zwave.commands.wakeupv2.WakeUpNotification cmd)** and **configure()**.
3. I will compare the configuration report I received to _**state."${getConfigurationInfo(idx, "name")}"**_ from the user has set. I will set _**state."${getConfigurationInfo(idx, "name")}Refresh"**_ to false if these are the same. Here are **zwaveEvent(physicalgraph.zwave.commands.configurationv2.ConfigurationReport cmd)** and **configurationCheck(config)**.
4. Loop from step 2 to step 3 until these configuration value are the same.

```
def updated() {
	configurationUpdate(settings)
	...
}

def configurationUpdate(config) {
	getConfigurationNumber().each { idx ->
		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
			if (state."${getConfigurationInfo(idx, "name")}" != config?."${getConfigurationInfo(idx, "name")}") {
				state."${getConfigurationInfo(idx, "name")}" = config?."${getConfigurationInfo(idx, "name")}"
				state."${getConfigurationInfo(idx, "name")}Refresh" = true
			}
		}
	}
}

def zwaveEvent(physicalgraph.zwave.commands.wakeupv2.WakeUpNotification cmd) {
	def cmds = []
	
	cmds += configure()
	if (cmds) {
		cmds << "delay 5000"
	}
	cmds << zwave.wakeUpV2.wakeUpNoMoreInformation().format()
	
	return cmds ? response(cmds) : []
}

def configure() {
	def cmds = []
	def value
	def optionValue
	...
	getConfigurationNumber().each { idx ->
		if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
			switch (getConfigurationInfo(idx, "type")) {
				case "enum":
					optionValue = getConfigurationInfo(idx, "enumMap").find { it.key == state."${getConfigurationInfo(idx, "name")}" }?.value
					value = integer2Array(optionValue, getConfigurationInfo(idx, "size"))
					break
				case "number":
					value = integer2Array(state."${getConfigurationInfo(idx, "name")}", getConfigurationInfo(idx, "size"))
					break
			}
			if (value != null) {
				cmds << zwave.configurationV2.configurationSet(parameterNumber: idx, defaultValue: false, size: getConfigurationInfo(idx, "size"), configurationValue: value).format()
				cmds << zwave.configurationV2.configurationGet(parameterNumber: idx).format()
				if (idx == 1) {
					cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x01, scale: optionValue?:0x00).format()
				}
				value = null
			}
		}
	}
	...
	return cmds ? delayBetween(cmds, 500) : []
}

def zwaveEvent(physicalgraph.zwave.commands.configurationv2.ConfigurationReport cmd) {
	def size = getConfigurationInfo(cmd.parameterNumber, "size")
	def value = array2Integer(cmd.configurationValue)
	def config = [:]
	
	if (size && size == cmd.size) {
		switch (getConfigurationInfo(cmd.parameterNumber, "type")) {
			case "enum":
				def optionName = getConfigurationInfo(cmd.parameterNumber, "enumMap").find { it.value == value}?.key
				if (optionName) {
					config."${getConfigurationInfo(cmd.parameterNumber, "name")}" = optionName
				}
				break
			case "number":
				config."${getConfigurationInfo(cmd.parameterNumber, "name")}" = value
				break
		}
		
		if (config) {
			configurationCheck(config)
		}
	}
	[]
}

def configurationCheck(config) {
	getConfigurationNumber().each { idx ->
		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
			if (config?."${getConfigurationInfo(idx, "name")}" == state."${getConfigurationInfo(idx, "name")}") {
				if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
					state."${getConfigurationInfo(idx, "name")}Refresh" = false
				}
			}
		}
	}
}
```

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Okay, I already use **scaledConfigurationValue** to replace my conversion functions. By the way, I can't see the **scaledConfigurationValue** in [class Configuration Report Command](https://graph.api.smartthings.com/ide/doc/zwave-utils.html#configurationV2), but **scaledConfigurationValue** still works fine in class Configuration Report Command.

File: devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy
Comment (greens) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: Okay so you are verifying they've been set.

My mistake, but this is just reinforcing my point that the code you've written is hard to follow because you're using integers combined with strings as your dictionary keys.

Iterating from 1..8 and checking to see if each param was in your updated configuration report is much more laborious and confusing than just structuring the output of your configuration report such that you could simply iterate over that itself.

There are other existing DTHs that solve this problem more elegantly: https://github.com/SmartThingsCommunity/SmartThingsPublic/blob/master/devicetypes/fibargroup/fibaro-dimmer-2-zw5.src/fibaro-dimmer-2-zw5.groovy for example.

Comment (vision-raytseng) [reply devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy]: I think I understand your meaning. Your approach simplifies this procedure and make it more elegantly.
I refactored configuration parameter using a map of maps, and refactored wake up info parameter using a map. Please review it. Thank you very much.

**Review by greens — APPROVED** [review]

**Comment by greens** [comment]
@KKlimczukS this now looks good to me, just waiting on your approval

**Review by KKlimczukS — APPROVED** [review]

**Pull request merged**
